core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -810,7 +810,7 @@ private[spark] object Utils extends Logging {
    * Get the local machine's URI.
    */
   def localHostNameForURI(): String = {
-    customHostname.getOrElse(InetAddresses.toUriString(localIpAddress))
+    sys.env.get("SPARK_PUBLIC_DNS").getOrElse( customHostname.getOrElse(InetAddresses.toUriString(localIpAddress)) )
```

Comment:
However, I'll double check this and, if it works reliably with `SPARK_LOCAL_HOSTNAME`, remove this bit.